dns/src/mdns.rs

```diff
@@ -0,0 +1,175 @@
+//! MDNS client for the [Wiznet W5500] SPI internet offload chip.
+
+use crate::header::{Header, Qr};
+#[cfg(not(feature = "std"))]
+use crate::ll::net::Eui48Addr;
+use crate::ll::{
+    net::{Ipv4Addr, SocketAddrV4},
+    Protocol, Sn, SocketCommand, SocketMode, SocketStatus,
+};
+use crate::{read_labels, Query, Response};
+use w5500_hl::{
+    io::{Read, Seek, SeekFrom},
+    Error, Hostname, Udp, UdpReader, UdpWriter,
+};
+
+const MDNS_PORT: u16 = 5353;
+const MDNS_ADDRESS: Ipv4Addr = Ipv4Addr::new(224, 0, 0, 251);
+const MDNS_SERVER: SocketAddrV4 = SocketAddrV4::new(MDNS_ADDRESS, MDNS_PORT);
+#[cfg(not(feature = "std"))]
+const MDNS_HARDWARE_DST: Eui48Addr = Eui48Addr::new(0x01, 0x00, 0x5E, 0x00, 0x00, 0xFB);
+
+/// W5500 MDNS client.
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct Client {
+    sn: Sn,
+    port: u16,
+    server: SocketAddrV4,
+}
+
+impl Client {
+    /// Create a new MDNS client.
+    ///
+    /// # Arguments
+    ///
+    /// * `sn` - The socket number to use for MDNS queries.
+    /// * `port` - MDNS source port.
+    ///   Use any free port greater than 1024 not in use by other W5500 sockets or None
+    ///   for the default MDNS port of 5353
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use w5500_dns::{ll::Sn, mdns::Client as MdnsClient};
+    ///
+    /// const MDNS_SRC_PORT: u16 = 45917;
+    ///
+    /// let mdns_client: MdnsClient = MdnsClient::new(Sn::Sn3, Some(MDNS_SRC_PORT));
+    /// ```
+    #[must_use]
+    pub fn new(sn: Sn, port: Option<u16>) -> Self {
+        Self {
+            sn,
+            port: port.unwrap_or(MDNS_PORT),
+            server: MDNS_SERVER,
+        }
+    }
+
+    /// A simple MDNS query.
+    ///
+    /// This will only broadcast an MDNS query, it will not wait for a reply.
+    fn query<'a, W5500: Udp>(
+        &mut self,
+        w5500: &'a mut W5500,
+    ) -> Result<Query<'a, W5500>, Error<W5500::Error>> {
+        // The register simulation doesn't implement either of
+        // these features, but they are necessary for proper
+        // MDNS operation.
+        #[cfg(not(feature = "std"))]
+        {
+            w5500.set_sn_dhar(self.sn, &MDNS_HARDWARE_DST)?;
+            w5500.set_sn_ttl(self.sn, 255)?;
+        }
```

Comment:
I added support for TTL: https://github.com/newAM/w5500-rs/pull/156

`dhar` is not implemented, but it does store the value, so it shouldn't panic (if it does let me know).
```suggestion
        w5500.set_sn_dhar(self.sn, &MDNS_HARDWARE_DST)?;
        w5500.set_sn_ttl(self.sn, 255)?;
```

Reply:
Unfortunately I get 

```
thread 'main' panicked at 'not yet implemented', regsim/src/lib.rs:987:31
```

Reply:
That one should go away if you merge `origin/main`